### PR TITLE
New resource: azurerm_stream_analytics_reference_input_blob

### PR DESCRIPTION
**Pull request by praneetloke**
I noticed that it was not possible to create inputs for a stream analytics job of type [reference](https://docs.microsoft.com/en-us/azure/stream-analytics/stream-analytics-use-reference-data#azure-blob-storage). The code to represent such a resource already existed, but just wasn't exposed as a resource to be able to use it.

I also added an example, that I tested using my own subscription. Here's a screenshot of how the "reference" input looks.

Screenshot_2019-06-10 Microsoft Azure

### CODE REVIEW

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
+package azurerm
```
Comment (praneetloke) [inline azurerm/resource_arm_stream_analytics_reference_input_blob.go]: This code is pretty much an exact replica of the `..._stream_analytics_**stream**_input_blob.go`.

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+	serializationRaw := d.Get("serialization").([]interface{})
+	serialization, err := azure.ExpandStreamAnalyticsStreamInputSerialization(serializationRaw)
+	if err != nil {
+		return fmt.Errorf("Error expanding `serialization`: %+v", err)
+	}
+
+	props := streamanalytics.Input{
+		Name: utils.String(name),
+		Properties: &streamanalytics.ReferenceInputProperties{
```
Comment (praneetloke) [inline azurerm/resource_arm_stream_analytics_reference_input_blob.go]: The main difference between the reference_input_blob and the stream_input_blob is this `properties` block.

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+			Type: streamanalytics.TypeReference,
```
Comment (praneetloke) [inline azurerm/resource_arm_stream_analytics_reference_input_blob.go]: This value is what sets it distinguishes a reference input from a streaming input to the stream analytics job.

File: azurerm/resource_arm_stream_analytics_reference_input_blob_test.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
+package azurerm
```
Comment (praneetloke) [inline azurerm/resource_arm_stream_analytics_reference_input_blob_test.go]: The tests in here are also pretty much a mirror of the "stream"_input_blob resource's tests.

File: examples/stream-analytics/main.tf [hunk 5]
```diff
@@ -0,0 +1,61 @@
+provider "azurerm" {}
```
Comment (praneetloke) [inline examples/stream-analytics/main.tf]: This is a working example. Although, I do have to clarify that I tested creating this entire plan, and also destroying it. I verified that the input for the job that is created is in fact of type "reference".

File: website/docs/r/stream_analytics_stream_input_blob.html.markdown [hunk 6]
```diff
@@ -37,8 +37,8 @@ resource "azurerm_storage_container" "example" {
   container_access_type = "private"
 }
 
-resource "azurerm_stream_analytics_stream_input_eventhub" "test" {
-  name                         = "eventhub-stream-input"
+resource "azurerm_stream_analytics_stream_input_blob" "test" {
```
Comment (praneetloke) [inline website/docs/r/stream_analytics_stream_input_blob.html.markdown]: Noticed the typos while reading the docs.

**Comment by praneetloke** [comment]
Ping. Anyone?

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go (hunk 2 above)
Comment (katbyte) [reply azurerm/resource_arm_stream_analytics_reference_input_blob.go]: Would it just make sense to add this to the other resource?

**Comment by katbyte** [comment]
As mentioned above, Would it just make sense to add this ability to the to the other steam input blob resource?

**Comment by praneetloke** [comment]
@katbyte I am going to take another look at this and get back to you. As you can see it has been a while since I opened this PR and have to refresh my memory a bit :)

**Comment by praneetloke** [comment]
> As mentioned above, Would it just make sense to add this ability to the to the other steam input blob resource?

Ok I looked at this again. I do believe this needs to be its own resource. Unfortunately, there is _some_ overlap in how _this_ resource is parsed vs. how the _stream_ input is parsed. But there are two places where we must explicitly use the _reference_ input type in the `*CreateUdpate` and the `*Read` functions. Maybe I don't know enough about adding a new resource that I can't see how this could be a part of the existing streaming input resource.

Just to be clear, it is important to note that the difference between the two types of input is that one if a _streaming_ input vs. what is being added with this PR is a _reference_ input. A _reference_ input is a static source of data for a Stream Analytics job, whereas a _streaming_ input is something that is continuously producing data.

Please let me know if this helps. I am happy to rework this if you have other ideas. I think for a Stream Analytics job, being able to add reference data inputs is an important feature.

**Comment by praneetloke** [comment]
ping @katbyte

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+			"date_format": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"path_pattern": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_stream_analytics_reference_input_blob.go]: Could we get a `validate.NoEmptyStrings` here?

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		if read.ID == nil {
+			return fmt.Errorf("Cannot read ID of Stream Analytics Reference Input Blob %q (Job %q / Resource Group %q)", name, jobName, resourceGroup)
+		}
+
+		d.SetId(*read.ID)
+	} else {
+		if _, err := client.Update(ctx, props, resourceGroup, jobName, name, ""); err != nil {
+			return fmt.Errorf("Error Updating Stream Analytics Reference Input Blob %q (Job %q / Resource Group %q): %+v", name, jobName, resourceGroup, err)
+		}
+	}
```
Comment (katbyte) [inline azurerm/resource_arm_stream_analytics_reference_input_blob.go]: Given there is vastly different logic for create & update could we maybe split this function into a create and update?

File: examples/stream-analytics/main.tf [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+resource "azurerm_storage_account" "example" {
+  name                     = "${var.prefix}-examplestoracc"
+  resource_group_name      = "${azurerm_resource_group.example.name}"
+  location                 = "${azurerm_resource_group.example.location}"
+  account_tier             = "Standard"
+  account_replication_type = "LRS"
+}
+
+resource "azurerm_storage_container" "example" {
+  name                  = "example"
```
Comment (katbyte) [inline examples/stream-analytics/main.tf]:
```suggestion
  name                  = "${var.prefix}example"
```

File: examples/stream-analytics/main.tf [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+  transformation_query = <<QUERY
+    SELECT *
+    INTO [YourOutputAlias]
+    FROM [YourInputAlias]
+QUERY
+}
+
+resource "azurerm_stream_analytics_reference_input_blob" "test" {
+  name                         = "blob-reference-input"
```
Comment (katbyte) [inline examples/stream-analytics/main.tf]:
```suggestion
  name                         = "${var.prefix}-blob-reference-input"
```

File: examples/stream-analytics/variables.tf [hunk 11]
```diff
@@ -0,0 +1,7 @@
+variable "prefix" {
+  description = "The prefix which should be used for all resources in this example"
+}
+
+variable "location" {
+  description = "The Azure Region in which all resources in this example should be created."
+}
```
Comment (katbyte) [inline examples/stream-analytics/variables.tf]: Could we put a logical default location here?

File: website/docs/r/stream_analytics_reference_input_blob.html.markdown [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+A `serialization` block supports the following:
+
+* `type` - (Required) The serialization format used for the reference data. Possible values are `Avro`, `Csv` and `Json`.
+
+* `encoding` - (Optional) The encoding of the incoming data in the case of input and the encoding of outgoing data in the case of output. It currently can only be set to `UTF8`.
+
+-> **NOTE:** This is required when `type` is set to `Csv` or `Json`.
+
+* `field_delimiter` - (Optional) The delimiter that will be used to separate comma-separated value (CSV) records. Possible values are ` ` (space), `,` (comma), `   ` (tab), `|` (pipe) and `;`.
```
Comment (katbyte) [inline website/docs/r/stream_analytics_reference_input_blob.html.markdown]:
```suggestion
* `field_delimiter` - (Optional) The delimiter that will be used to separate comma-separated value (CSV) records. Possible values are ` ` (space), `,` (comma), `	` (tab), `|` (pipe) and `;`.
```

File: website/docs/r/stream_analytics_reference_input_blob.html.markdown [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+* `resource_group_name` - (Required) The name of the Resource Group where the Stream Analytics Job exists. Changing this forces a new resource to be created.
+
+* `stream_analytics_job_name` - (Required) The name of the Stream Analytics Job. Changing this forces a new resource to be created.
+
+* `date_format` - (Required) The date format. Wherever `{date}` appears in `path_pattern`, the value of this property is used as the date format instead.
+
+* `path_pattern` - (Required) The blob path pattern. Not a regular expression. It represents a pattern against which blob names will be matched to determine whether or not they should be included as input or output to the job.
+
+* `storage_account_name` - (Required) The name of the Storage Account.
```
Comment (katbyte) [inline website/docs/r/stream_analytics_reference_input_blob.html.markdown]: The name of the strage account to.. ?

File: examples/stream-analytics/variables.tf (hunk 11 above)
Comment (praneetloke) [reply examples/stream-analytics/variables.tf]: I am happy to, but I was following the other examples in this provider which don't seem to set a default location. If you would like me to still set it, I am fine with it.

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go (hunk 7 above)
Comment (praneetloke) [reply azurerm/resource_arm_stream_analytics_reference_input_blob.go]: Done.

Comment (praneetloke) [reply azurerm/resource_arm_stream_analytics_reference_input_blob.go]: Just curious, though. With `Required: true`, isn't `NoEmptyStrings` enforced automatically?

**Comment by danielrbradley** [comment]
Hello, I'm personally very interested in getting Reference sources working for a project. I've read through the conversation so far and can't quite work out what the current blocker would be (beyond fixing merge conflicts). I'd be happy to try and lend a hand if I can to get this over the line and merged if you can summarise the outstanding issues that need to be resolved.

**Comment by praneetloke** [comment]
@danielrbradley 

> beyond fixing merge conflicts

I believe this and also getting the test to pass. I haven't kept my fork up-to-date with the latest from the upstream master. I'll get this restored today.

**Comment by danielrbradley** [comment]
@praneetloke @katbyte I've rebased and got the build working on this second PR: #5129

Quite happy @praneetloke if you'd like to pull my changes and force push up to your original branch to continue the conversation here.

**Comment by praneetloke** [comment]
@danielrbradley I am taking a look at restoring this PR right now. Thanks for opening yours. I'll check that too. I would prefer to fix up this PR rather than opening a new one, since the comments in this PR would be easier to follow-up on.

**Comment by praneetloke** [comment]
@katbyte @tombuildsstuff I have addressed the comments and fixed-up this PR with the latest from upstream (this repo). Can you please take another look at this?

**Comment by praneetloke** [comment]
Almost all of the checks passed, except for the `lintstatic` target, which seems to have failed with a transient error:

```
# cd .; git clone -- https://github.com/client9/misspell /home/travis/gopath/src/github.com/client9/misspell
Cloning into '/home/travis/gopath/src/github.com/client9/misspell'...
fatal: unable to access 'https://github.com/client9/misspell/': Failed to connect to github.com port 443: Connection timed out
package github.com/client9/misspell/cmd/misspell: exit status 128
GNUmakefile:13: recipe for target 'tools' failed
```

I don't have access to restart the build. Could you please restart just [this](https://travis-ci.org/terraform-providers/terraform-provider-azurerm/jobs/624282159) job?

File: azurerm/resource_arm_stream_analytics_reference_input_blob.go (hunk 7 above)
Comment (katbyte) [reply azurerm/resource_arm_stream_analytics_reference_input_blob.go]: Nope, you can pass an empty string, alls that is required is something is assinged to the property

Comment (praneetloke) [reply azurerm/resource_arm_stream_analytics_reference_input_blob.go]: Ah gotcha. Would you like me to add `validate.NoEmptyStrings` here then?

**Pull request merged**